### PR TITLE
ci: solidity test benchmarks

### PR DESCRIPTION
**Pull request by agostbiro**
This PR adds automated regression checks for the JS Solidity test runner. The setup is similar to how the RPC benchmark that benchmarks Hardhat Node-style workloads works.

## Benchmark Suite

The benchmark runs the test harness of `forge-std`. This has two advantages:
- Largest coverage of cheatcodes in a single test harness which is helpful to catch regressions in all features.
- Makes sure our implementation works with `forge-std` which we intend to use.
 
The disadvantage is that `forge-std` is not representative of how users are writing Solidity tests. But this is not a concern as long as we’re not using the benchmark to guide optimizations, only to catch regressions.

There are multiple benchmark scenarios. The first one is called `Total` which measures the entire test harness execution of the `forge-std` test harness. The purpose of this scenario is to make sure that there are no regressions in the parallel test suite executor.

In addition to test suites, individual tests within test suites are also parallelized. This means that if we want to catch regressions in features covered by a certain test suite, we have to execute that test suite in isolation. For this reason, in addition to the `Total` scenario, we execute the following test suites as well:

- `StdCheatsTest`: [environment](https://book.getfoundry.sh/cheatcodes/environment)-related cheatcode coverage + fuzzing.
- `StdStorageTest`: storage-related cheatcode coverage + fuzzing.
- `StdMathTest`: math-related utilities + it's very quick (~15ms on my machine), so regressions from FFI overhead should show up on this.
- `StdUtilsForkTest`: read-type programmatic fork tests.
- `StdCheatsForkTest`: write + fuzz tests for programmatic forking.

The rest of the test suites in `forge-std` test things implemented in Solidity or test file-system utilities, so they're less interesting for us.

## Benchmark Metric

The metric for all scenarios is the wall clock time measured from JS in milliseconds. The measurement excludes loading files from disk as that's out of scope for EDR. If there is a 10% regression on any of the scenarios above, the benchmark job fails (same as the RPC benchmark).

## Baseline 

The baseline is executing the same tests with `forge test`. Instructions to run the baseline are in the [crates/tools/js/benchmark/src/solidity-tests.ts](https://github.com/NomicFoundation/edr/blob/e5eb6201ff6fede9373e08fc87c7e90ffb77878b/crates/tools/js/benchmark/src/solidity-tests.ts) file. 

I tested manually that performance is identical to `forge test` on my machine, so we don't have to keep running the baseline.

## Benchmark Jobs

There are two benchmark workflows in the `feat/solidity-tests` branch and three benchmark jobs:

- `EDR Benchmarks`:
  - `Run JS scenario runner benchmark for Hardhat Node style workload`
    - Stores results for `main` and compares PRs against against those results
    - Dashboard: https://nomic-foundation-automation.github.io/edr-benchmark-results/bench/
  - `Run JS Solidity test runner benchmark`
    - This is the new job
    - Stores results for `feat/solidity-tests` and compares PRs targeting `feat/solidity-tests` against those results
      - Once `feat/solidity-tests` is merged to `main`, we'll switch to storing results for `main`
    - Dashboard will be live here once this PR is merged: https://nomic-foundation-automation.github.io/edr-benchmark-results/soltests/

- `EDR Benchmarks for feat/solidity-tests`
  - `Run JS scenario runner benchmark for Hardhat Node style workload`
    - Stores results for `feat/solidity-tests` and compares PRs against against those results
    - Once `feat/solidity-tests` is merged to `main`, we'll remove this

I reorganized the workflow files a bit to share Rust compilation cache + start caching the RPC cache for the Solidity test runner benchmarks.

### CODE REVIEW

File: crates/tools/js/benchmark/src/solidity-tests.ts
Comment (Wodann) [inline crates/tools/js/benchmark/src/solidity-tests.ts:180]: Maybe add a comment here to indicate that we are using a seed to ensure deterministic execution.

File: .github/workflows/edr-benchmark-feat-solidity-tests.yml
Comment (Wodann) [inline .github/workflows/edr-benchmark-feat-solidity-tests.yml]: `main` looks like it might be changed by mistake

**Review by Wodann — APPROVED** [review]
Overall looks good. There is one erroneous change that you might want to look into before merging. The other comment is a suggestion

File: crates/tools/js/benchmark/src/solidity-tests.ts
Comment (agostbiro) [reply crates/tools/js/benchmark/src/solidity-tests.ts:181]: Thanks, fixed in [08dacd2](https://github.com/NomicFoundation/edr/pull/686/commits/08dacd20152819aa9b765ffca7af96865aff498f)

**Comment by agostbiro** [comment]
There is a failing benchmark indicating a performance regression compared to `main`: https://github.com/NomicFoundation/edr/actions/runs/11121119547/job/30899561763?pr=686

This is not actually a regression, what happened was that EDR got faster in `main` and those changes are not yet in this branch as merging `main` into `feat/solidity-tests` is blocked by independent JS package resolution issues.

File: .github/workflows/edr-benchmark-feat-solidity-tests.yml
Comment (agostbiro) [reply .github/workflows/edr-benchmark-feat-solidity-tests.yml]: Apologies, I got confused a bit. It actually needs to be `feat/solidity-tests` here to make it run in the branch, but we want to change it back to `main` before we merge `feat/solidity-tests` into main. Fixed it here with a TODO comment: [Fix precondition for soltests benchmark](https://github.com/NomicFoundation/edr/pull/524/commits/ae8625fa3457eb5d4bd1f402959881bc91c2015f)

**Pull request merged**
